### DISCUSSION TITLE
Define styles for variant + size combination

### DISCUSSION THREAD

**Discussion by dlindenkreuz** [chakra-ui/chakra-ui#4492] [Ideas]
Currently, Chakra allows to define styles for sizes and variants that can be combined independently. In some cases, I feel there is a need to also define what happens when a variant is used together with a certain size.

**Simple Example: Button "circular" variant**
```js
// ... theme definition
Button: {
  variants: {
    circular: {
      borderRadius: "full",
      bg: "whiteAlpha.100",
      // I suggest to add the following:
      sizes: {
        md: {
          minW: 0,
          boxSize: 8
        }
      }
    },
  },
}
```

**Issue:**
With the [default size definitions](https://github.com/chakra-ui/chakra-ui/blob/main/packages/theme/src/components/button.ts#L164-L189), this button would not have a circular shape, but assume the (fixed) height of its respective size.

**Solution:**
Add a size definition that applies only to the `circular` variant.

Attributes of a size class that are defined in the context of one variant would override attributes defined by the "generic" size class.

(I know that one could also achieve this through component composition, and I have heard of the `<Circle/>` component 😉 The example above would be an alternative.)

Any thoughts? 🤔

**Comment by anubra266** [comment]
Yous should define a size variant on its own, then call it in the variant
```js
// ... theme definition
Button: {
  variants: {
    circular: {
      borderRadius: "full",
      bg: "whiteAlpha.100", 
      size: "new_md"
    },
  },
  sizes: {
        new_md: {
          minW: 0,
          boxSize: 8
        }
      }
}
```

**Comment by Midori-307** [comment]
I stumbled over the same problem and noticed that if you use the function notation (as described here: https://chakra-ui.com/docs/styled-system/theming/advanced), you can use conditional styles as the size is passed with the `StyleFunctionProps`:
```tsx

// ...

const getCircularStyle = ({size}: StyleFunctionProps) => {
  const sizeProps = size === "md" ? {
    minW: 0,
    boxSize: 8
  } : undefined;

  return {
    ...sizeProps,
    borderRadius: "full",
    bg: "whiteAlpha.100",
  };
 };

export const Button = {
  baseStyle,
  sizes,
  variants: { circular: getCircularStyle },
  defaultProps,
};
```
